### ISSUE TITLE
Consider removing the @JsonCreator requirement

### ISSUE THREAD

**Issue by lpandzic** [FasterXML/jackson-module-parameter-names#13]
This feature would really simplify usage and provide convience for developers.
Considering everyday development techniques I've encountered regarding this issue the most basic use case boils down to:

```java

class AClass {

  ... mandatory field m1, ..., mn
  ... optional field o1, ..., on

  AClass(m1, ..., mn) {
  }

  ... setters for o1, .., on

  ... other methods
}

```

One additional argument for tackling this issue are value types which [will always be immutable](http://cr.openjdk.java.net/~briangoetz/valhalla/specialization.html).

**Comment by cowtowncoder** [comment]
I have an idea which would apply to `jackson-databind`, and might satisfy different requirements (convenience of detecting likely primary constructor; safety of not accidentally using a constructor that was not mean to be used). Given that without a no-arg constructor, `@JsonCreator`, or auto-detected single-argument (String, boolean, int/long, double) constructor there is no way to deserialize a value, I think it would be safe to auto-detect a multi-argument constructors if:

1. It is the only visible constructor class has
2. There are no `@JsonCreator` annotated factory methods
3. All arguments have either a name (implicit or explicit), or are injectable

Would this make sense?

/cc @christophercurrie

**Comment by lpandzic** [comment]
> Would this make sense?

Yes.
I've made a pull request with tests for this issue.

**Comment by cowtowncoder** [comment]
Ok, one more related improvement:

https://github.com/FasterXML/jackson-databind/issues/725

which I hope should actually further reduce the need for `@JsonCreator`. Fix is in master (for 2.6.0-SNAPSHOT). The basic rule is simple: if

1. There is no other visible creator and
2. All parameters have explicit or implicit name

a constructor will be used similar to if it was annotated. Cases that do not fall under this would be:

a. Multiple visible constructors: no attempt is made to know which one to use; no point guessing as there will always be some cases where determination impossible (same number of arguments, same visibility etc)
b. Single-argument case: due to ambiguity between delegating/properties-based variants this case does require annotation (default is to assume creator to be delegating)

**Comment by lpandzic** [comment]
Nice, so after the release the tests should pass.

**Comment by lpandzic** [comment]
I've tried it out locally and tests pass with jackson-databind 2.6.0-SNAPSHOT!
After release README.md should be updated to reflect those changes - min. dependency and removal of @JsonCreator in the example.

**Comment by lpandzic** [comment]
This issue is solved so I'm closing it.
Maybe somewhere in the future we can come up with some smarter exact algorithm for cases of multiple visible constructors.

**Comment by cowtowncoder** [comment]
Agreed.

**Comment by lpandzic** [comment]
Is release planned soon?

**Comment by cowtowncoder** [comment]
No immediate plans, but I do hope to start release planning once JsonParser-filtering is implemented for streaming API. The other bigger thing (protobuf dataformat) is complete.
We may be able to start Release Candidate process before end of April.

**Comment by lpandzic** [comment]
Does this latest release contain this feature? (jackson-databind)

**Comment by cowtowncoder** [comment]
2.6.0 has not yet been released, last one was 2.5.3, so not yet.
